### PR TITLE
[oneTBB] Add adaptive mutexes

### PR DESCRIPTION
**Pull request by pavelkumbrasev**
Signed-off-by: pavelkumbrasev <pavel.kumbrasev@intel.com>

### CODE REVIEW

**Comment by pavelkumbrasev** [comment]
@alexey-katranov @ivankochin @alexandraepan

File: source/elements/oneTBB/source/mutual_exclusion/mutex_cls.rst [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+..
+.. SPDX-License-Identifier: CC-BY-4.0
+
+=====
+mutex
+=====
+**[mutex.mutex]**
+
+A ``mutex`` is a class that models the :doc:`Mutex requirement <../named_requirements/mutexes/mutex>`,
+using adaptive approach: the combination of spinlock and waiting on system primitives.
```
[the remark this answers is not in the document]
Comment (pavelkumbrasev) [reply source/elements/oneTBB/source/mutual_exclusion/mutex_cls.rst:11]:
```suggestion
using an adaptive approach, which is a combination of spinlock and waiting on system primitives.
```

File: source/elements/oneTBB/source/mutual_exclusion/rw_mutex_cls.rst [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+..
+.. SPDX-License-Identifier: CC-BY-4.0
+
+=============
+rw_mutex
+=============
+**[mutex.rw_mutex]**
+
+A ``rw_mutex`` is a class that models the :doc:`ReaderWriterMutex requirement <../named_requirements/mutexes/rw_mutex>`,
+using adaptive approach: the combination of spinlock and waiting on system primitives.
```
[the remark this answers is not in the document]
Comment (pavelkumbrasev) [reply source/elements/oneTBB/source/mutual_exclusion/rw_mutex_cls.rst:11]:
```suggestion
using an adaptive approach, which is a combination of spinlock and waiting on system primitives.
```

**Comment by ivankochin** [comment]
LGTM!

File: source/elements/oneTBB/source/mutual_exclusion/mutex_cls.rst [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+..
+.. SPDX-License-Identifier: CC-BY-4.0
+
+=====
+mutex
+=====
+**[mutex.mutex]**
+
+A ``mutex`` is a class that models :doc:`Mutex requirement <../named_requirements/mutexes/mutex>`
+using an adaptive approach, which is a combination of spinlock and waiting on system primitives.
```
Comment (pavelkumbrasev) [inline source/elements/oneTBB/source/mutual_exclusion/mutex_cls.rst:11]:
```suggestion
using an adaptive approach, it guarantees: that the thread that can't acquire the lock spins before blocking.
```

File: source/elements/oneTBB/source/mutual_exclusion/rw_mutex_cls.rst [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+..
+.. SPDX-License-Identifier: CC-BY-4.0
+
+=============
+rw_mutex
+=============
+**[mutex.rw_mutex]**
+
+A ``rw_mutex`` is a class that models :doc:`ReaderWriterMutex requirement <../named_requirements/mutexes/rw_mutex>`
+using an adaptive approach, which is a combination of spinlock and waiting on system primitives.
```
Comment (pavelkumbrasev) [inline source/elements/oneTBB/source/mutual_exclusion/rw_mutex_cls.rst:11]:
```suggestion
using an adaptive approach, it guarantees: that the thread that can't acquire the lock spins before blocking.
```

File: source/elements/oneTBB/source/mutual_exclusion/mutex_cls.rst [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+..
+.. SPDX-License-Identifier: CC-BY-4.0
+
+=====
+mutex
+=====
+**[mutex.mutex]**
+
+A ``mutex`` is a class that models :doc:`Mutex requirement <../named_requirements/mutexes/mutex>`
+using an adaptive approach, it guarantees: that the thread that can't acquire the lock spins before blocking.
```
Comment (alexandraepan) [inline source/elements/oneTBB/source/mutual_exclusion/mutex_cls.rst:11]:
```suggestion
using an adaptive approach, it guarantees that the thread that cannot acquire the lock spins before blocking.
```

File: source/elements/oneTBB/source/mutual_exclusion/mutex_cls.rst [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+.. cpp:function:: ~mutex()
+
+    Destroys an unlocked ``mutex``.
+
+--------------------------------------------------
+
+.. cpp:function:: void lock()
+
+    Acquires a lock. It uses adaptive logic for waiting, thus it is blocked after a certain time of busy waiting.
```
Comment (alexandraepan) [inline source/elements/oneTBB/source/mutual_exclusion/mutex_cls.rst:68]:
```suggestion
    Acquires a lock. It uses an adaptive logic for waiting, thus it is blocked after a certain time of busy waiting.
```

File: source/elements/oneTBB/source/mutual_exclusion/rw_mutex_cls.rst [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+..
+.. SPDX-License-Identifier: CC-BY-4.0
+
+=============
+rw_mutex
+=============
+**[mutex.rw_mutex]**
+
+A ``rw_mutex`` is a class that models :doc:`ReaderWriterMutex requirement <../named_requirements/mutexes/rw_mutex>`
+using an adaptive approach, it guarantees: that the thread that can't acquire the lock spins before blocking.
```
Comment (alexandraepan) [inline source/elements/oneTBB/source/mutual_exclusion/rw_mutex_cls.rst:11]:
```suggestion
using an adaptive approach, it guarantees that the thread that cannot acquire the lock spins before blocking.
```

**Review by alexandraepan — APPROVED** [review]

**Review by alexey-katranov — APPROVED** [review]

File: source/elements/oneTBB/source/mutual_exclusion/mutex_cls.rst [hunk 8]
```diff
@@ -0,0 +1,80 @@
+.. SPDX-FileCopyrightText: 2020-2021 Intel Corporation
```
Comment (kboyarinov) [inline source/elements/oneTBB/source/mutual_exclusion/mutex_cls.rst:1]:
```suggestion
.. SPDX-FileCopyrightText: 2021 Intel Corporation
```

File: source/elements/oneTBB/source/mutual_exclusion/rw_mutex_cls.rst [hunk 9]
```diff
@@ -0,0 +1,104 @@
+.. SPDX-FileCopyrightText: 2020-2021 Intel Corporation
```
Comment (kboyarinov) [inline source/elements/oneTBB/source/mutual_exclusion/rw_mutex_cls.rst:1]:
```suggestion
.. SPDX-FileCopyrightText: 2021 Intel Corporation
```

**Review by kboyarinov — CHANGES_REQUESTED** [review]

**Review by kboyarinov — APPROVED** [review]

**Pull request merged**